internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1072,86 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
+
+	snapshots, err := a.loader.GetAllSnapshotsForPR(a.context, a.client, application, snapshot.GetLabels()[gitops.PipelineAsCodePullRequestAnnotation], snapshot.GetLabels()[gitops.PipelineAsCodeRepoURLAnnotation])
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the application",
+			"application.Name:", application.Name)
+		return err
+	}
+	sortedSnapshots := gitops.SortSnapshots(*snapshots)
+	if len(sortedSnapshots) < 1 {
+		a.logger.Info("There are no found snapshots for given repo at the moment.")
+		return nil
+	}
+	// check for older snapshots, their current status and type
+	for i := 1; i < len(sortedSnapshots); i++ {
+		if gitops.IsComponentSnapshot(&sortedSnapshots[i]) {
+			err = gitops.MarkSnapshotAsCanceled(a.context, a.client, &sortedSnapshots[i], "Snapshot canceled/superseded")
+			if err != nil {
+				a.logger.Error(err, "Failed to mark snapshot as canceled", "snapshot.Name", snapshot.Name)
+				return err
+			}
+			// get all integration pipelineruns for a snapshot
+			integrationTestPipelineruns, err := getAllIntegrationPipelineRunsForSnapshot(a.context, a.client, &sortedSnapshots[i])
+			if err != nil {
+				a.logger.Error(err, "Failed to get all integration pipelineruns for snapshot", "snapshot.Name", snapshot.Name)
+				return err
+			}
+			sortedPlrs := gitops.SortPipelineruns(integrationTestPipelineruns)
+			if len(sortedPlrs) < 1 {
+				a.logger.Info("There are no pipelineruns for the snapshot at the moment.")
+				return nil
+			}
+			latestPlrName := sortedPlrs[0].Name
+			for _, plr := range integrationTestPipelineruns {
+				plr := plr
+				if plr.Name == latestPlrName {
+					continue
+				}
+				pipelinerunStatus, _, err := h.GetRunningIntegrationPipelineRunStatus(a.context, a.client, &plr)
+				if err != nil {
+					a.logger.Error(err, "Failed to get integration test pipelinerun status", "snapshot.Name", plr.Name)
+					return err
+				}
+				if pipelinerunStatus == intgteststat.IntegrationTestStatusInProgress {
+					//remove finalizer and cancel pipelinerun
+					err = h.RemoveFinalizerFromPipelineRun(a.context, a.client, a.logger, &plr, h.IntegrationPipelineRunFinalizer)
+					if err != nil {
+						a.logger.Error(err, "Failed to remove finalizer from integration test pipelinerun")
+						return err
+					}
+					// set "CancelledRunFinally" to PLR status, should gracefully cancel pipelinerun, this is so raw I hate this
+					patch := client.MergeFrom(plr.DeepCopy())
+					plr.Spec.Status = tektonv1.PipelineRunSpecStatusCancelledRunFinally
+					err := a.client.Patch(a.context, &plr, patch)
+					if err != nil {
+						return err
+					}
+
+				}
+			}
+
+		}
+	}
+	return err
+}
+
+func getAllIntegrationPipelineRunsForSnapshot(ctx context.Context, adapterClient client.Client, snapshot *applicationapiv1alpha1.Snapshot) ([]tektonv1.PipelineRun, error) {
```

Comment:
Not sure but I'm thinking if should this function be moved to loader?

Reply:
Agreed, moved to loader.